### PR TITLE
feat: dev server can now reuse the same tab on Mac

### PR DESCRIPTION
**Pull request by ExiaSR**
Inspire by CRA, `gatsby develop -o` can now reuse the same tab on MacOS with Google Chrome.

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit cceaa3933b88369bfd53a0e94d673e95db1631e7

https://app.netlify.com/sites/using-postcss-sass/deploys/5b4fa6184ed62f31bac85d62

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit e1fbc0bcb2eab96c7057bc54d5a17b5bf3ac22a5

https://app.netlify.com/sites/using-postcss-sass/deploys/5b4fa68267610c36a9a0ec69

**Comment by m-allanson** [comment]
Woah, neat! Thanks @ExiaSR. 

Although this adds a fair bit of code for a  (common, but) very specific OS and browser combination. I wonder if a better idea would be to try and add this feature directly to [the `opn` package](https://github.com/sindresorhus/opn)? Then any project could make use of it, instead of it needing to be implemented by every project that wants to reuse a Chrome tab.

**Comment by ExiaSR** [comment]
@m-allanson I think reuseing the same tab is really a web development specific thing, e.g. http://localhost:8000 and opn isn’t just for opening an URL in browser, it also supprts stuffs like image, text file and more options.

**Comment by ExiaSR** [comment]
@m-allanson Any update?

**Comment by pieh** [comment]
How about creating and publishing new package for it that would fallback to `opn`, so it's abstracted from gatsby and could be used in other projects? Then we could "just" switch `opn` to that package?

This feature defenitely improves developer experience and would love to have it. It's just I'm not sure if we can maintain it here in the long run

**Comment by ExiaSR** [comment]
@pieh sounds good to me :)
I’ll open a new PR when it’s done.

**Comment by ExiaSR** [comment]
- [better-opn](https://www.npmjs.com/package/better-opn)

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbyjs* failed.

Built with commit 4b4564aff44f0e5bbe5d30768aeba3f2e56ae815

https://app.netlify.com/sites/gatsbyjs/deploys/5b5897e2fdd72a1eee8b3740

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbyjs* failed.

Built with commit 4a81c953c1a605cb2aaf9d87053e3415a56adc7e

https://app.netlify.com/sites/gatsbyjs/deploys/5b589871c6aed61a298d5213

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbyjs* failed.

Built with commit 48008d1f83be88aea6b1a3777df9b5b67f04180c

https://app.netlify.com/sites/gatsbyjs/deploys/5b5899fcb31274665a1cc192

**Comment by piotrkwiecinski** [comment]
On linux chrome always opens new tab.

**Comment by ExiaSR** [comment]
Well, as I mentioned in the description. It only works for Chrome on MacOS.

Sent from my iPhone

On Jul. 31, 2018, at 3:05 PM, Piotr Kwiecinski <notifications@github.com<mailto:notifications@github.com>> wrote:


On linux chrome always opens new tab.

**Comment by KyleAMathews** [comment]
Deploy preview for *image-processing* failed.

Built with commit 366be585184b05f153ede09a5080fa9d229d901e

https://app.netlify.com/sites/image-processing/deploys/5b800bbac6aed6630959add9

**Comment by ExiaSR** [comment]
@pieh Do you guys still consider merging this? It's been a while.

Also, some help with the `e2e_tests_path-prefix` will be really appreciated. I tried to run the test locally and it worked.

**Comment by DSchau** [comment]
@michaek we're still ironing out the e2e tests. Need to merge a few PRs before the end of the week so we can get in better shape re: those.

So don't worry too much about e2e test failures! I think it's unlikely this PR would introduce any regressions re: e2e.

**Comment by ExiaSR** [comment]
@DSchau Is there any blocker for this PR? or it just has a lower priority?

File: packages/gatsby/package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
     "babel-plugin-dynamic-import-node": "^1.2.0",
     "babel-plugin-macros": "^2.4.0",
     "babel-plugin-remove-graphql-queries": "^2.0.2-rc.3",
+    "better-opn": "^0.1.0",
```
Comment (DSchau) [inline packages/gatsby/package.json]: Is there value in keeping `opn` or should we just always use `better-opn`?

Looks like it's used in the serve task as well.

File: packages/gatsby-cli/README.md [hunk 2]
```diff
@@ -37,6 +37,8 @@ development server.
 | `-o`, `--open`  | Open the site in your (default) browser for you |
 | `-S`, `--https` | Use HTTPS                                       |
 
+With `-o` flag enabled, for MacOS Chrome user, we will try our best to reuse the same tab. For any other scenarios, it will fallback to open a new tab instead.
```
Comment (DSchau) [inline packages/gatsby-cli/README.md]: I would just remove this, perhaps. It's almost not worth documenting the "edge" case, and the dev will just get that experience when using?

Comment (ExiaSR) [reply packages/gatsby-cli/README.md]: @DSchau Make sense to me, I guess most people won't even read this file.

File: packages/gatsby/package.json (hunk 1 above)
Comment (ExiaSR) [reply packages/gatsby/package.json]: @DSchau We should always use `better-opn`. However, I make `opn` as a peer dependency at `better-opn` (check [here[(https://github.com/ExiaSR/better-opn/blob/master/package.json#L32), `opn` still have to be explicitly added as a dependency.

**Comment by pieh** [comment]
Hey, I checked and couple of things - right now it errors out:
```
error UNHANDLED EXCEPTION


  TypeError: require(...)(...).catch is not a function

  - develop.js:413 compiler.hooks.done.tapAsync
    [strict-mode]/[gatsby]/dist/commands/develop.js:413:63
```

I guess that comes from the fact that you did change API a bit - returning boolean true/false instead of promise as in `opn`. You also silence rejected promises in `better-opn` ( https://github.com/ExiaSR/better-opn/blob/2258a6ea579d10a175283d8b2ec71e4c4431eefb/index.js#L66 ). So we would never actually show that message.

Additionally `better-opn` doesn't work currently on node 6 (which we support):
```sh
➜  better-opn-test node -v
v6.14.4
➜  better-opn-test node index.js
/Users/misiek/dev/better-opn-test/node_modules/better-opn/index.js:64
    var options = { app: browser, ...opts };
                                  ^^^

SyntaxError: Unexpected token ...
```




And other thing we should update https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/commands/serve.js as well for consistency

**Comment by ExiaSR** [comment]
@pieh 
Hi! `better-opn` now keeps the API consistent with `opn` which returning a promise instead boolean https://github.com/ExiaSR/better-opn/commit/8e189d99514093d1834d51b2dbd895bf14a0f5c8. Also, it nows support Node 6 https://github.com/ExiaSR/better-opn/commit/91d498f9e52dd8a66cd478d1df359b2533f72810.

**Comment by pieh** [comment]
There were situation where promise was not returned - for example with `BROWSER=none gatsby develop -o`, `better-opn` was returning false instead of promise, so I wrapped call in Promise.resolve to not break things. I also pinned `better-opn` to current version as it's in pre 1.0 version to prevent any unintentional breakage.

**Pull request merged**